Tools/TestWebKitAPI/Tests/WebKitCocoa/RestoreLocalStorage.mm

```diff
@@ -50,32 +60,76 @@ static void testFetchLocalStorage(RetainPtr<WKWebsiteDataStore> websiteDataStore
     [webView loadHTMLString:@"<body>webkit.org</body>" baseURL:[NSURL URLWithString:@"https://webkit.org"]];
     [navigationDelegate waitForDidFinishNavigation];
 
+    // Put data into local storage.
     __block bool doneEvaluatingJavaScript = false;
     [webView evaluateJavaScript:@"localStorage.setItem('key', 'value');" completionHandler:^(id value, NSError *error) {
         EXPECT_NULL(error);
         doneEvaluatingJavaScript = true;
     }];
     TestWebKitAPI::Util::run(&doneEvaluatingJavaScript);
 
+    // Fetch the local storage data.
     RetainPtr set = [NSSet setWithObject:WKWebsiteDataTypeLocalStorage];
 
     __block bool doneFetching = false;
+    __block RetainPtr<NSData> localStorageData;
     [[[webView configuration] websiteDataStore] _fetchDataOfTypes:set.get() completionHandler:^(NSData *data) {
         EXPECT_NOT_NULL(data);
         EXPECT_GT([data length], 0u);
+        localStorageData = data;
         doneFetching = true;
     }];
     TestWebKitAPI::Util::run(&doneFetching);
+
+    // Restore Local Storage.
+
+    // Kill network and web processes.
+    [websiteDataStore _terminateNetworkProcess];
+    [webView _killWebContentProcessAndResetState];
+
+    // Clear the data store.
+    doneRemoving = false;
+    [websiteDataStore removeDataOfTypes:[WKWebsiteDataStore allWebsiteDataTypes] modifiedSince:[NSDate distantPast] completionHandler:^{
+        doneRemoving = true;
+    }];
+    TestWebKitAPI::Util::run(&doneRemoving);
+
+    // Create a new WebView.
+    RetainPtr newViewConfiguration = adoptNS([WKWebViewConfiguration new]);
+    [newViewConfiguration setWebsiteDataStore:websiteDataStore.get()];
+    RetainPtr newWebView = adoptNS([[WKWebView alloc] initWithFrame:CGRectMake(0, 0, 100, 100) configuration:newViewConfiguration.get()]);
+
+    // Restore the local storage data.
+    [[[newWebView configuration] websiteDataStore] _restoreDataOfTypes:set.get() data:localStorageData.get()];
+
+    WTF::sleep(1_s);
+
+    // Load the same page as before.
+    RetainPtr newNavigationDelegate = adoptNS([[TestNavigationDelegate alloc] init]);
+    [newWebView setNavigationDelegate:newNavigationDelegate.get()];
+    [newWebView loadHTMLString:@"<body>webkit.org</body>" baseURL:[NSURL URLWithString:@"https://webkit.org"]];
+    [newNavigationDelegate waitForDidFinishNavigation];
+
+    // Check that the local storage data was correctly restored.
+    doneEvaluatingJavaScript = false;
+    [newWebView evaluateJavaScript:@"localStorage.getItem('key');" completionHandler:^(id value, NSError *error) {
+        EXPECT_NULL(error);
+        // This is currently failing (value is null).
```

Comment:
Progress report: I think the overall structure of the patch there. But it's currently failing :(. Will debug more tomorrow.

Reply:
It works now :)